metamorphic (race): failure from 5/19

Found via a metamorphic test run ([logs](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/5202092?buildTab=log&focusLine=5281&linesState=4494)).

```console
===== SEED =====
1652935592018725775
===== ERR =====
exit status 66
===== OUT =====
==================
WARNING: DATA RACE
Write at 0x00c000233b18 by goroutine 25:
  runtime.slicecopy()
      GOROOT/src/runtime/slice.go:284 +0x0
  github.com/cockroachdb/pebble/sstable.(*blockIter).readEntry()
      github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/block.go:416 +0xb7e
  github.com/cockroachdb/pebble/sstable.(*blockIter).First()
      github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/block.go:793 +0xfd
  github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).firstInternal()
      github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader.go:791 +0x111
  github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).First()
      github.com/cockroachdb/pebble/sstable/external/com_github_cockroachdb_pebble/sstable/reader.go:768 +0x64
  github.com/cockroachdb/pebble.(*levelIter).skipEmptyFileForward()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_iter.go:621 +0x74
  github.com/cockroachdb/pebble.(*levelIter).Next()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_iter.go:560 +0x3a8
  github.com/cockroachdb/pebble.(*simpleMergingIter).step()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_checker.go:226 +0x2431
  github.com/cockroachdb/pebble.checkLevelsInternal()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_checker.go:660 +0x1629
  github.com/cockroachdb/pebble.(*DB).CheckLevels()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/level_checker.go:576 +0x264
  github.com/cockroachdb/pebble.DebugCheckLevels()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/options.go:736 +0x32
  github.com/cockroachdb/pebble/internal/metamorphic.(*test).init.func1.1()
      github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/test.go:58 +0x33
  github.com/cockroachdb/pebble/internal/metamorphic.withRetries()
      github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/retryable.go:18 +0x2e
  github.com/cockroachdb/pebble/internal/metamorphic.(*test).init.func1()
      github.com/cockroachdb/pebble/internal/metamorphic/external/com_github_cockroachdb_pebble/internal/metamorphic/test.go:57 +0x4a
  github.com/cockroachdb/pebble.(*DB).updateReadStateLocked()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/read_state.go:105 +0x33b
  github.com/cockroachdb/pebble.(*DB).compact1()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1916 +0xc0c
  github.com/cockroachdb/pebble.(*DB).compact.func1()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1840 +0xe4
  runtime/pprof.Do()
      GOROOT/src/runtime/pprof/runtime.go:40 +0x125
  github.com/cockroachdb/pebble.(*DB).compact()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1837 +0xb5
  github.com/cockroachdb/pebble.(*DB).maybeScheduleCompactionPicker·dwrap·13()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1631 +0x58

Previous read at 0x00c000233b1e by goroutine 7:
  [failed to restore the stack]

Goroutine 25 (running) created at:
  github.com/cockroachdb/pebble.(*DB).maybeScheduleCompactionPicker()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1631 +0x1044
  github.com/cockroachdb/pebble.(*DB).maybeScheduleCompaction()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1530 +0x1a9
  github.com/cockroachdb/pebble.(*DB).compact.func1()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1847 +0x19d
  runtime/pprof.Do()
      GOROOT/src/runtime/pprof/runtime.go:40 +0x125
  github.com/cockroachdb/pebble.(*DB).compact()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1837 +0xb5
  github.com/cockroachdb/pebble.(*DB).maybeScheduleCompactionPicker·dwrap·13()
      github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/compaction.go:1631 +0x58

Goroutine 7 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:1306 +0x726
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:1598 +0x99
  testing.tRunner()
      GOROOT/src/testing/testing.go:1259 +0x22f
  testing.runTests()
      GOROOT/src/testing/testing.go:1596 +0x7ca
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:1504 +0x9d1
  main.main()
      main/bazel-out/k8-dbg/bin/external/com_github_cockroachdb_pebble/internal/metamorphic/metamorphic_test_/testmain.go:131 +0x5c4
==================
```

SHA: `9bd431430a38`.

_NOTE: I filled this in manually, as we didn't get an automated issue._

With [some instrumentation](https://github.com/cockroachdb/pebble/commit/cee350804d6dd5602f049541d144f84bc9ca2564) that tracks `fullKey` buffer pointers of iterators in the pool, I confirmed the `mergingIter` is still holding on to and reading a key from closed sstable reader. In the below execution, iterator `0xc001dab500` with buffer `0xc0002dfb20` is returned to the pool, and the mergingIter reads the buffer `0xc0002dfb20` shortly later. 
 
```
Close returned sstable iterator 0xc001c50e00 with fullKey buffer 0xc00080c528 to the pool.
heapDebugContext seekPrefix after init min heap (mergingIter = 0xc0015c0e20)
github.com/cockroachdb/pebble.(*mergingIter).heapDebugContext:928
github.com/cockroachdb/pebble.(*mergingIter).seekGE:923
github.com/cockroachdb/pebble.(*mergingIter).isNextEntryDeleted:642
github.com/cockroachdb/pebble.(*mergingIter).findNextEntry:662
github.com/cockroachdb/pebble.(*mergingIter).SeekPrefixGE:983

Close returned sstable iterator 0xc0019b3880 with fullKey buffer 0xc001d12fa8 to the pool.
heapDebugContext SeekPrefixGE (mergingIter = 0xc0015c0e20)
github.com/cockroachdb/pebble.(*mergingIter).heapDebugContext:928
runtime.gopanic:844
github.com/cockroachdb/pebble.(*mergingIter).findNextEntry:674
github.com/cockroachdb/pebble.(*mergingIter).SeekPrefixGE:983
github.com/cockroachdb/pebble/internal/keyspan.(*InterleavingIter).SeekPrefixGE:200

Close returned sstable iterator 0xc001dab500 with fullKey buffer 0xc0002dfb20 to the pool.
NewIter acquired sstable iterator 0xc001dab500 with fullKey buffer 0xc0002dfb20
--- FAIL: TestMeta (2.74s)
panic: mergingIter 0xc0015c0e20 attempting to read key 0xc00080c528 ("uwomy@5") [recovered]
	panic: mergingIter 0xc0015c0e20 attempting to read key 0xc00080c528 ("uwomy@5")
```

--- a/merging_iter.go
+++ b/merging_iter.go
@@ -649,27 +649,35 @@ func (m *mergingIter) isNextEntryDeleted(item *mergingIterItem) bool {
 
 // Starting from the current entry, finds the first (next) entry that can be returned.
 func (m *mergingIter) findNextEntry() (*InternalKey, []byte) {
+	var reseeked bool
 	for m.heap.len() > 0 && m.err == nil {
 		item := &m.heap.items[0]
 		if m.levels[item.index].isSyntheticIterBoundsKey {
 			break
 		}
+		// For prefix iteration, stop if we already seeked the iterator due to a
+		// range tombstone and are now past the prefix. We could amortize the
+		// cost of this comparison, by doing it only after we have iterated in
+		// this for loop a few times. But unless we find a performance benefit
+		// to that, we do the simple thing and compare each time. Note that
+		// isNextEntryDeleted already did at least 4 key comparisons in order to
+		// return true, and additionally at least one heap comparison to step to
+		// the next entry.
+		//
+		// Note that we cannot move this comparison into the isNextEntryDeleted
+		// branch. Once isNextEntryDeleted determines a key is deleted and seeks
+		// the level's iterator, item.key's memory is potentially invalid. If
+		// the iterator is now exhausted, item.key may be garbage.
+		if m.prefix != nil && reseeked {
+			if n := m.split(item.key.UserKey); !bytes.Equal(m.prefix, item.key.UserKey[:n]) {
+				return nil, nil
+			}
+		}
+
 		m.addItemStats(item)
 		if m.isNextEntryDeleted(item) {
 			m.stats.PointsCoveredByRangeTombstones++
-			// For prefix iteration, stop if we are past the prefix. We could
-			// amortize the cost of this comparison, by doing it only after we
-			// have iterated in this for loop a few times. But unless we find
-			// a performance benefit to that, we do the simple thing and
-			// compare each time. Note that isNextEntryDeleted already did at
-			// least 4 key comparisons in order to return true, and
-			// additionally at least one heap comparison to step to the next
-			// entry.
-			if m.prefix != nil {
-				if n := m.split(item.key.UserKey); !bytes.Equal(m.prefix, item.key.UserKey[:n]) {
-					return nil, nil
-				}
-			}
+			reseeked = true
 			continue
 		}
 		if item.key.Visible(m.snapshot) &&

